module/VuFind/src/VuFind/Related/Bookplate.php

```diff
@@ -0,0 +1,114 @@
+<?php
+/**
+ * Related Records: Solr-based similarity
+ *
+ * PHP version 7
+ *
+ * Copyright (C) Villanova University 2009.
+ *
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 2,
+ * as published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301  USA
+ *
+ * @category VuFind
+ * @package  Related_Records
+ * @author   Demian Katz <demian.katz@villanova.edu>
+ * @license  http://opensource.org/licenses/gpl-2.0.php GNU General Public License
+ * @link     https://vufind.org/wiki/development:plugins:related_records_modules Wiki
+ */
+namespace VuFind\Related;
+
+/**
+ * Related Records: Solr-based similarity
+ *
+ * @category VuFind
+ * @package  Related_Records
+ * @author   Demian Katz <demian.katz@villanova.edu>
+ * @license  http://opensource.org/licenses/gpl-2.0.php GNU General Public License
+ * @link     https://vufind.org/wiki/development:plugins:related_records_modules Wiki
+ */
+class Bookplate implements RelatedInterface
+{
+    /**
+     * Bookplate strings
+     */
+    protected $bookplateStrs;
+
+    /**
+     * Bookplate URLs
+     */
+    protected $bookplateImgNames;
+
+    /**
+     * URL template for full bookplate
+     */
+    protected $fullUrlTemplate;
+
+    /**
+     * URL temlate for thumbnail
+     */
+    protected $thumbUrlTemplate;
+
+    /**
+     * Display bookplate titles?
+     */
+    protected $displayTitles;
+
+    /**
+     * Establishes base settings for bookplates.
+     *
+     * @param string                            $settings Settings from config.ini
```

Comment:
As alluded in an earlier comment, I think it would be useful to leverage $settings for some or all of the module's configuration. Any values attached to the module name in config.ini will be passed here as the $settings parameter, so, for example, if I configured "related[] = Bookplate:foo:bar" then $settings would be set to "foo:bar" here.

The advantage to this is that you can configure multiple instances of Bookplate with different behaviors, and you can turn this on in multiple backends, and each instance can have its own configuration. If you hard-code the configuration paths, you reduce flexibility.

There are two options for how to take advantage of this:

1.) Put all the settings directly into the Related declaration, in a delimited form. Given the large number of config options you are proposing, I don't recommend this, as it would be unwieldy and unreadable.

2.) Pass in the names of a configuration file and section -- i.e. "Related:config:Record" -- and pull settings from there. You just need to create a factory that injects the `VuFind\Config\PluginManager` service to the module's constructor so that you have the ability to load configuration files as needed... then the init() method can pull the appropriate configuration and set everything up. (And, of course, if $settings is empty, you can default to config and Record to keep the existing default behavior).

I clearly have a preference for option 2!

Reply:
@demiankatz I don't think I quite know how to create a factory that injects the `VuFind\Config\PluginManager` service to the module's constructor and access the settings. I created a `BookplateFactory` which is executing but I don't really know where to take it from there. Could you give me some more detail or point me to a similar example? Sorry for the hassle.

Reply:
Sure, no problem! In the factory, you just need to pull the config manager out of the container and pass it to the object being built, e.g.:

<pre>
$configManager = $container->get(\VuFind\Config\PluginManager::class);
return new $requestedName($configManager);
</pre>

Then you can assign that to a property in Bookplate::__construct(), and go from there.

Does that help? I'm happy to dive deeper as needed.

Reply:
Thanks @demiankatz! That seems to have gotten me over the hump. Much appreciated!

Reply:
Fantastic, glad I could help! It's definitely a bit unintuitive at first, but once you get the hang of it, the Laminas factory system is pretty pleasant to work with. (At least, it gives me a certain weird "virtual Lego" satisfaction). :-)

Reply:
I think I have this setup correctly using option 2. I'm not sure if I got any of the details wrong but it's generally setup this way now.